test/avro_json_encoder_canon_tests.erl

```diff
@@ -0,0 +1,161 @@
+%%%   http://www.apache.org/licenses/LICENSE-2.0
+%%%
+%%% Unless required by applicable law or agreed to in writing,
+%%% software distributed under the License is distributed on an
+%%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%%% KIND, either express or implied.  See the License for the
+%%% specific language governing permissions and limitations
+%%% under the License.
+%%%-------------------------------------------------------------------
+-module(avro_json_encoder_canon_tests).
+
+-include_lib("eunit/include/eunit.hrl").
+
+canon(Json) ->
+  Schema = avro_json_decoder:decode_schema(Json),
+  avro:encode_schema(Schema, #{canon => true}).
+
+% Run Java test cases from the Avro project:
+% https://github.com/apache/avro/blob/master/share/test/data/schema-tests.txt
+
+java_primitive_test() ->
+  ?assertEqual(<<"\"null\"">>, canon(<<"\"null\"">>)),                % 000
+  ?assertEqual(<<"\"null\"">>, canon(<<"{\"type\":\"null\"}">>)),     % 001
+  ?assertEqual(<<"\"boolean\"">>, canon(<<"\"boolean\"">>)),          % 002
+  ?assertEqual(<<"\"boolean\"">>, canon(<<"{\"type\":\"boolean\"}">>)), % 003
+  ?assertEqual(<<"\"int\"">>, canon(<<"\"int\"">>)),                  % 004
+  ?assertEqual(<<"\"int\"">>, canon(<<"{\"type\":\"int\"}">>)),       % 005
+  ?assertEqual(<<"\"long\"">>,canon(<<"\"long\"">>)),                 % 006
+  ?assertEqual(<<"\"long\"">>, canon(<<"{\"type\":\"long\"}">>)),     % 007
+  ?assertEqual(<<"\"float\"">>, canon(<<"\"float\"">>)),              % 008
+  ?assertEqual(<<"\"float\"">>, canon(<<"{\"type\":\"float\"}">>)),   % 009
+  ?assertEqual(<<"\"double\"">>, canon(<<"\"double\"">>)),            % 010
+  ?assertEqual(<<"\"double\"">>, canon(<<"{\"type\":\"double\"}">>)), % 011
+  ?assertEqual(<<"\"bytes\"">>, canon(<<"\"bytes\"">>)),              % 012
+  ?assertEqual(<<"\"bytes\"">>, canon(<<"{\"type\":\"bytes\"}">>)),   % 013
+  ?assertEqual(<<"\"string\"">>, canon(<<"\"string\"">>)),            % 014
+  ?assertEqual(<<"\"string\"">>, canon(<<"{\"type\":\"string\"}">>)), % 015
+  % avro_json_decoder:decode_schema/1 considers empty enum invalid
+  % ?assertEqual(<<"[]">>, canon(<<"[  ]">>)),                        % 016
+  ?assertEqual(<<"[\"int\"]">>, canon(<<"[ \"int\"  ]">>)),           % 017
+  ?assertEqual(<<"[\"int\",\"boolean\"]">>,
+               canon(<<"[ \"int\" , {\"type\":\"boolean\"} ]">>)).    % 018
+
+% Put fields in standard order, without whitespace
+java_019_test() ->
+  ?assertEqual(<<"{\"name\":\"foo\",\"type\":\"record\",\"fields\":[]}">>,
+               canon(<<"{\"fields\":[], "
+                       "\"type\":\"record\", \"name\":\"foo\"}">>)).
+
+java_020_test() ->
+  ?assertEqual(<<"{\"name\":\"x.y.foo\",\"type\":\"record\",\"fields\":[]}">>,
+               canon(<<"{\"fields\":[], \"type\":\"record\", \"name\":\"foo\", "
+                       "\"namespace\":\"x.y\"}">>)).
+
+java_021_test() ->
+%  % The schema parser (avro_record:type/3) requires that the namespace must
```

Comment:
Please help to fix in avro_record.erl

Reply:
fixed in #61

Reply:
Updated to use upstream test, which now passes.